### PR TITLE
Tickets/dm 20212

### PR DESCRIPTION
**Pull request by jgates108**

### CODE REVIEW

File: admin/tools/docker/4_build-configured-images.sh [hunk 1]
```diff
@@ -89,6 +89,47 @@ if [ "$PUSH_TO_HUB" = "true" ]; then
     printf "Image %s pushed successfully\n" "$TAG"
 fi
 
+# Build the master-multi image
+
+awk \
+-v NODE_TYPE="-c" \
```
Comment (andy-slac) [inline admin/tools/docker/4_build-configured-images.sh]: Could you indent continuation lines, would be easier to read.

File: admin/tools/docker/configured/scripts/configure.sh [hunk 2]
```diff
@@ -44,12 +46,16 @@ EOD
 NODE_TYPE="worker"
 
 # get the options
-while getopts hm c ; do
-    case $c in
+while getopts ":hmcs" opt ; do
+    echo "Config option ${opt}"
```
Comment (andy-slac) [inline admin/tools/docker/configured/scripts/configure.sh]: That looks like debug printout that should not be used in normal running.

File: admin/tools/docker/configured/scripts/configure.sh [hunk 3]
```diff
@@ -44,12 +46,16 @@ EOD
...
+    case ${opt} in
             h) usage ; exit 0 ;;
             m) NODE_TYPE="master" ;;
+            c) NODE_TYPE="master-multi" ;;
+            s) NODE_TYPE="master-shared" ;;
             \?) usage ; exit 2 ;;
     esac
+    echo "Config option ${opt} node ${NODE_TYPE}"
```
Comment (andy-slac) [inline admin/tools/docker/configured/scripts/configure.sh]: This too

File: admin/tools/docker/configured/scripts/start.sh [hunk 4]
```diff
@@ -71,4 +72,11 @@ fi
 # TODO: wmgr.secret should be retrieved at container execution time
 cp $QSERV_RUN_DIR/etc/wmgr.secret.example $QSERV_RUN_DIR/etc/wmgr.secret
 
+# Wait for configure script to run.
+while [ ! -f "$QSERV_RUN_DIR"/configure_done ]
+do
+    echo "start.sh - wait for configure to finish on $(hostname)"
+    sleep 1
+done
```
Comment (andy-slac) [inline admin/tools/docker/configured/scripts/start.sh]: I'm not sure why do you have to loop and wait here, the code above runs in foreground?

File: admin/tools/docker/configured/scripts/wait.sh [hunk 5]
```diff
@@ -29,8 +29,16 @@ set -e
...
+# Wait for configure script to run.
+while [ ! -f "$QSERV_RUN_DIR"/configure_done ]
+do
+    echo "Wait for configure to finish on $(hostname)"
+    sleep 2
+done
+
 # Wait for Qserv services to be up and running
-while ! "$QSERV_RUN_DIR"/bin/qserv-status.sh > /dev/null
+#while ! "$QSERV_RUN_DIR"/bin/qserv-status.sh > /dev/null
```
Comment (andy-slac) [inline admin/tools/docker/configured/scripts/wait.sh]: REmove this line if it's not needed

File: admin/tools/docker/deployment/localhost/run-multiczar-tests.sh [hunk 6]
```diff
@@ -0,0 +1,110 @@
+#!/bin/sh
+
+# @author  Fabrice Jammes, IN2P3/SLAC
```
Comment (andy-slac) [inline admin/tools/docker/deployment/localhost/run-multiczar-tests.sh]: Did Fabrice wrote this?
I think we should drop all `@author` strings from code, git history is better for that purpose.

File: admin/tools/docker/deployment/localhost/run-multiczar-tests.sh [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+. "${DIR}/multiczar-env.sh"
+
+usage() {
+  cat << EOD
+
+  Usage: $(basename "$0") [options]
+
+  Available options:
+    -h          this message
+    -C          Rebuild the images from scratch
```
Comment (andy-slac) [inline admin/tools/docker/deployment/localhost/run-multiczar-tests.sh]: I don't see `-C` option handled below.

File: admin/tools/docker/deployment/localhost/run-multiczar-tests.sh [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+docker rm -f "$MASTER" || echo "No existing container for $MASTER"
+docker run --detach=true \
+    --privileged \
+    --cap-add sys_admin \
+    --cap-add sys_ptrace \
+    -v /home/jgates/work/qserv_testdata:/qserv/qserv_testdata \
+    -e "QSERV_MASTER=$MASTER" --name "$MASTER" -h "${MASTER}" "$MASTER_SHARED_IMAGE"
+#    -v /home/jgates/qserv/data-qserv:/qserv/data/qserv \
+#    -v /home/jgates/work/qserv_testdata:/qserv/qserv_testdata \
+#    -e "QSERV_MASTER=$MASTER" --name "$MASTER" -h "${MASTER}" "$MASTER_SHARED_IMAGE"
```
Comment (andy-slac) [inline admin/tools/docker/deployment/localhost/run-multiczar-tests.sh]: Commented code should be removed if not needed.

File: admin/tools/docker/deployment/localhost/run-multiczar-tests.sh [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+MASTER_IP=$(docker inspect -f '{{ .NetworkSettings.IPAddress }}' $MASTER)
+HOSTFILE="${HOSTFILE}$MASTER_IP  master.localdomain
+"
```
Comment (andy-slac) [inline admin/tools/docker/deployment/localhost/run-multiczar-tests.sh]: Quotes on separate line look odd

File: core/modules/qproc/DatabaseModels.h [hunk 10]
```diff
@@ -0,0 +1,74 @@
+/*
+ * LSST Data Management System
+ * Copyright 2019 AURA/LSST.
```
Comment (andy-slac) [inline core/modules/qproc/DatabaseModels.h]: Not AURA

File: core/modules/qproc/DatabaseModels.h [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+    DatabaseModels& operator=(DatabaseModels const&) = delete;
+
+    virtual ~DatabaseModels() = default;
+
+    static Ptr create(std::map<std::string, std::string> const& config);
+    static Ptr create(sql::SqlConfig const& cfg);
+
+    bool applySql(std::string const& sql, sql::SqlResults& results, sql::SqlErrorObject& errObj);
+
+    std::vector<std::string> listColumns(std::string const& dbName, std::string const& tableName);
```
Comment (andy-slac) [inline core/modules/qproc/DatabaseModels.h]: Should this method be `const`?
Need comments for every method.

File: core/modules/qproc/DatabaseModels.h [hunk 12]
```diff
@@ -0,0 +1,74 @@
...
+
+private:
+    //explicit DatabaseModels(mysql::MySqlConfig const& mySqlConfig);
```
Comment (andy-slac) [inline core/modules/qproc/DatabaseModels.h]: This is not a comment 🙂

File: core/modules/qproc/DatabaseModels.cc [hunk 13]
```diff
@@ -0,0 +1,96 @@
+/*
+ * LSST Data Management System
+ * Copyright 2019 AURA/LSST.
```
Comment (andy-slac) [inline core/modules/qproc/DatabaseModels.cc]: Not AURA.

File: core/modules/qproc/DatabaseModels.cc [hunk 14]
```diff
@@ -0,0 +1,96 @@
...
+LOG_LOGGER _log = LOG_GET("lsst.qserv.qproc.DatabaseModels");
+}
+
+namespace lsst {
+namespace qserv {
+namespace qproc {
+
+DatabaseModels::Ptr DatabaseModels::create(map<string, string> const& configMap) {
+    util::ConfigStore cfgStore(configMap);
+    /// Use the CSS config for now. The CSS database is not used but sql::SqlConnection wont work without one.
```
Comment (andy-slac) [inline core/modules/qproc/DatabaseModels.cc]: I'm not sure how this comment is relevant to the code below?

File: core/modules/rproc/InfileMerger.cc [hunk 15]
```diff
@@ -1,7 +1,7 @@
 // -*- LSST-C++ -*-
 /*
  * LSST Data Management System
- * Copyright 2014-2016 AURA/LSST.
+ * Copyright 2014-2019 AURA/LSST.
```
Comment (andy-slac) [inline core/modules/rproc/InfileMerger.cc]: Not AURA

File: core/modules/wmgr/python/dbMgr.py [hunk 16]
```diff
@@ -201,7 +201,7 @@ def createDb():
         # grant full access to non-privileged account
         cmd = "GRANT ALL PRIVILEGES ON {0}.* TO '{1}'@'{2}'"
         user = Config.instance().dbUser
-        hostnames = ['localhost']
+        hostnames = ['localhost', '%%']
```
Comment (andy-slac) [inline core/modules/wmgr/python/dbMgr.py]: Why is there two percent signs here?

File: admin/templates/installation/qserv-meta.conf [hunk 17]
```diff
@@ -52,7 +52,7 @@ node_type = mono
 # Qserv master DNS name
 # WARNING : for mono-node install, do not use value localhost,
 # but 127.0.0.1 instead
-master = 127.0.0.1
+master = master.localdomain
```
[the remark this answers is not in the document]
Comment (jgates108) [reply admin/templates/installation/qserv-meta.conf]: Only mono needs the master.localdomain entry. The multinode and multiczar setup addresses in their scripts.

File: admin/templates/configuration/etc/my-proxy.cnf [hunk 18]
```diff
@@ -3,8 +3,7 @@
 proxy-address = :{{MYSQLPROXY_PORT}}
 
 # proxy-backend-addresses = 127.0.0.1:3306
-# TODO : is it usefull for multi node ?
-proxy-backend-addresses = {{MYSQLD_HOST}}:{{MYSQLD_PORT}}
+proxy-backend-addresses = {{MYSQLDSHARED_HOST}}:{{MYSQLDSHARED_PORT}}
```
[the remark this answers is not in the document]
Comment (jgates108) [reply admin/templates/configuration/etc/my-proxy.cnf]: I think we will absolutely need a central location (source of truth) for the schema of installed databases. It is expected that user will be able to load their own data and without a central location to keep this straight, there will be problems.

File: admin/templates/configuration/etc/xrdssi.cnf [hunk 19]
```diff
@@ -2,9 +2,8 @@
 # Default values for parameters are commented
 
 [mysql]
-
-# hostname =
-# port =
+hostname = master.localdomain
```
[the remark this answers is not in the document]
Comment (jgates108) [reply admin/templates/configuration/etc/xrdssi.cnf]: Mono setup works but you have to add an entry for master.localdomain in /etc/hosts. multinode and multiczar work without it.

File: admin/tools/docker/configured/scripts/start.sh (hunk 4 above)
Comment (jgates108) [reply admin/tools/docker/configured/scripts/start.sh]: Once it starts the container, it moves on to the next line. It doesn't wait for the container, so wait.sh would try to run qserv-status.sh before the container is done with qserv-configure.py, which is really confusing and probably not safe either. Waiting for the qserv-configure.py to finish forces things to run in a consistent order.

File: admin/templates/installation/qserv-meta.conf (hunk 17 above)
[the remark this answers is not in the document]
Comment (andy-slac) [reply admin/templates/installation/qserv-meta.conf]: I don't think it is acceptable. We need to be able to run tests with mono setup without root. We should probably make this host name configurable and use different values depending on mono- or multi-czar setup.

File: admin/templates/configuration/etc/my-proxy.cnf (hunk 18 above)
[the remark this answers is not in the document]
Comment (andy-slac) [reply admin/templates/configuration/etc/my-proxy.cnf]: Proxy's backend has to be a database with result tables, if you connect proxy to a shared instance how do you get results?

File: admin/templates/configuration/etc/my-client.cnf [hunk 20]
```diff
@@ -2,6 +2,6 @@
 [client]
 user     = {{MYSQLD_USER_QSERV}}
 # host/port and/or socket
-host     = {{MYSQLD_HOST}}
-port     = {{MYSQLD_PORT}}
+host     = {{MYSQLDSHARED_HOST}}
+port     = {{MYSQLDSHARED_PORT}}
```
[the remark this answers is not in the document]
Comment (jgates108) [reply admin/templates/configuration/etc/my-client.cnf]: It looks like it is being used by qserv-configure.py.

File: core/modules/wmgr/python/dbMgr.py (hunk 16 above)
Comment (jgates108) [reply core/modules/wmgr/python/dbMgr.py]: It doesn't work with only one.

File: admin/templates/installation/qserv-meta.conf (hunk 17 above)
[the remark this answers is not in the document]
Comment (jgates108) [reply admin/templates/installation/qserv-meta.conf]: I modified configure.py to overwrite master and shared with 127.0.0.1 when in mono node, so the /etc/hosts entry isn't needed anymore.

File: core/modules/qmeta/QMeta.h [hunk 21]
```diff
@@ -86,7 +86,7 @@ class QMeta {
     virtual ~QMeta() {}
 
     /**
-     *  @brief Return czar ID given czar "name".
+     *  @brief Return czar ID given czar "name" (host:.
```
Comment (jgates108) [inline core/modules/qmeta/QMeta.h]: removed.

File: core/modules/qmeta/QMetaMysql.cc [hunk 22]
```diff
@@ -142,6 +142,7 @@ QMetaMysql::getCzarID(std::string const& name) {
     }
 }
 
+
```
Comment (jgates108) [inline core/modules/qmeta/QMetaMysql.cc]: removed

File: admin/python/lsst/qserv/admin/dataLoader.py [hunk 23]
```diff
@@ -82,12 +82,15 @@ class DataLoader(object):
     def __init__(self, configFiles, czarWmgr, workerWmgrMap={}, chunksDir="./loader_chunks",
                  chunkPrefix='chunk', keepChunks=False, skipPart=False, oneTable=False,
                  css=None, cssClear=False, indexDb='qservMeta', tmpDir=None,
-                 emptyChunks=None, deleteTables=False, loggerName=None, cssDb='qservCssData'):
+                 emptyChunks=None, deleteTables=False, loggerName=None, cssDb='qservCssData',
+                 czarWmgrList=[]):
```
Comment (andy-slac) [inline admin/python/lsst/qserv/admin/dataLoader.py]: In Python it is not recommended to use modifiable object (like list or dict) for a default value of parameters. Better way is to use `None` and then test for it in the code.

File: admin/python/lsst/qserv/admin/dataLoader.py [hunk 24]
```diff
@@ -82,12 +82,15 @@ class DataLoader(object):
...
         """
         Constructor parses all arguments and prepares for execution.
 
         @param configFiles:  Sequence of the files defining all partitioning options.
-        @param czarWmgr:     WmgrClient instance for czar node.
+        @param czarWmgr:     WmgrClient instance for czar/master node.
```
Comment (andy-slac) [inline admin/python/lsst/qserv/admin/dataLoader.py]: My guess here is that "czar/master" does not run czar in multi-czar mode, it is only for mysql (and cmsd)? This whole naming is a bit confusing I wish we had some distinction for what is what.

File: admin/templates/configuration/etc/init.d/mysql-proxy [hunk 25]
```diff
@@ -77,6 +86,7 @@ start_cmd() {
     start -p $pidfile -l $lockfile $NAME "${cmd}"
     ret=$?
     [ $ret -eq 0 ] || echo "See startup logfile : ${logfile}"
+    prime_qserv
```
Comment (andy-slac) [inline admin/templates/configuration/etc/init.d/mysql-proxy]: I can imagine that this could fail occasionally if proxy is slow to start for some reason or if it does not bind mysql socket immediately.

File: admin/tools/docker/4_build-configured-images.sh [hunk 26]
```diff
@@ -89,17 +89,58 @@ if [ "$PUSH_TO_HUB" = "true" ]; then
...
+
+TAG="${DOCKER_IMAGE}_master_shared"
+printf "Building master image %s from %s\n" "$TAG" "$DOCKERDIR"
+docker build --tag="$TAG" "$DOCKERDIR"
+printf "Image %s built successfully\n" "$TAG"
+
+if [ "$PUSH_TO_HUB" = "true" ]; then
+    docker push "$TAG"
+    printf "Image %s pushed successfully\n" "$TAG"
+fi
```
Comment (andy-slac) [inline admin/tools/docker/4_build-configured-images.sh]: Lots of repeated code here, maybe move that into a function to avoid repeating?

File: core/modules/ccontrol/UserQueryResources.cc [hunk 27]
```diff
@@ -38,14 +38,17 @@ UserQuerySharedResources::UserQuerySharedResources(std::shared_ptr<css::CssAcces
...
         : css(css_),
         mysqlResultConfig(mysqlResultConfig_),
         secondaryIndex(secondaryIndex_),
         queryMetadata(queryMetadata_),
         queryStatsData(queryStatsData_),
         qMetaSelect(qMetaSelect_),
-        resultDbConn(resultDbConn_)
+        resultDbConn(resultDbConn_),
+        databaseModels(dbModels_)
+
```
Comment (andy-slac) [inline core/modules/ccontrol/UserQueryResources.cc]: extra empty line

File: core/modules/qmeta/QMeta.h (hunk 21 above)
Comment (andy-slac) [reply core/modules/qmeta/QMeta.h]: I can still see it :)

File: core/modules/wcontrol/Foreman.cc [hunk 28]
```diff
@@ -102,6 +102,8 @@ void Foreman::processTask(std::shared_ptr<wbase::Task> const& task) {
             auto qr = wdb::QueryRunner::newQueryRunner(task, _chunkResourceMgr, _mySqlConfig);
             qr->runQuery();
         }
+        // Transmission is done, but 'task' contains statistics that are still useful.
+        task->sendChannel.reset(); // Frees its xrdsvc::SsiRequest object.
```
Comment (andy-slac) [inline core/modules/wcontrol/Foreman.cc]: This does not seem to be related to multi-czar?

File: admin/python/lsst/qserv/admin/dataLoader.py (hunk 24 above)
Comment (jgates108) [reply admin/python/lsst/qserv/admin/dataLoader.py]: It's the czar when when in mono mode or multinode, and the master when in multiczar. I think multinode may eventually go away as multiczar runs all of the same tests on multiple nodes, but mono is still handy as it's a lot faster to run and easier to debug. It could use a better name.

File: admin/templates/configuration/etc/init.d/mysql-proxy (hunk 25 above)
Comment (jgates108) [reply admin/templates/configuration/etc/init.d/mysql-proxy]: It's an ugly hack, I agree. I'll be happy to see it go, but this ticket's scope is much too big already, and poking around the proxy to get qserv to start right away can wait a little. I don't think start returns until after the proxy is really running.

I don't think this hack is really needed at this point, but its existence encourages someone (anyone) to change the proxy. If there's consensus, I'll remove it before merging.

File: core/modules/mysql/MySqlConnection.h [hunk 29]
```diff
@@ -85,6 +85,8 @@ class MySqlConnection : boost::noncopyable {
     MySqlConfig const& getConfig() const { return *_sqlConfig; }
     bool selectDb(std::string const& dbName);
 
+    std::string dump();
```
Comment (n8pease) [inline core/modules/mysql/MySqlConnection.h]: needs a docstring

File: core/modules/qproc/DatabaseModels.h [hunk 30]
```diff
@@ -0,0 +1,95 @@
...
+public:
+    using Ptr = std::shared_ptr<DatabaseModels>;
+
+    DatabaseModels() = delete;
+    DatabaseModels(DatabaseModels const&) = delete;
+    DatabaseModels& operator=(DatabaseModels const&) = delete;
+
+    virtual ~DatabaseModels() = default;
+
+    /// @return a DatabaseModels object from config (see util::configStor)
```
Comment (n8pease) [inline core/modules/qproc/DatabaseModels.h]: configStor -> ConfigStore?

Comment (n8pease) [reply core/modules/qproc/DatabaseModels.h]: I don't see any changes to configStor(e) in this changelist. Anyway, please add some description of the use of 'master' and 'local' here.

File: core/modules/qproc/QuerySession.cc [hunk 31]
```diff
@@ -298,13 +299,21 @@ void QuerySession::finalize() {
...
+    : _css(t.css), _defaultDb(t.defaultDb) {
+    sql::SqlConfig sqlConfig(t.sqlConfig);
+    _databaseModels = DatabaseModels::create(sqlConfig, sqlConfig);
     _initContext();
 }
 
 
+QuerySession::QuerySession() {
+    sql::SqlConfig sqlConfig(sql::SqlConfig::MOCK);
+    _databaseModels = DatabaseModels::create(sqlConfig, sqlConfig);
```
Comment (n8pease) [inline core/modules/qproc/QuerySession.cc]: passing in the config twice is an odd pattern. Maybe an overload taking only one config instance is appropriate?

Comment (jgates108) [reply core/modules/qproc/QuerySession.cc]: Same config for both local and master, since local is also master. I'm not sure adding another create statement helps but I'll add a comment.
